DPI-Aware GetSystemMetrics and wxSystemParametersInfo

MaartenBent opened the pull request:
Cherry-picked from #334.

Because TLW's don't have a DPI yet, use the DPI from `GetDeviceCaps(hdc, LOGPIXELSY);` as input for `GetSystemMetricsForDpi` and `SystemParametersInfoForDpi`.

MaartenBent committed:
Add DPI Aware replacement of GetSystemMetrics

Use GetSystemMetricsForDpi when this function is available and the DPI is
known. Otherwise keep using GetSystemMetrics.

MaartenBent committed:
Use DPI Aware wxGetSystemMetrics

If no wxWindow is known, use wxTheApp->GetTopWindow().
Also use a wxWindow for all wxSystemSettings::GetMetric calls.

MaartenBent committed:
Add DPI Aware replacement of SystemParametersInfo

Use SystemParametersInfoForDpi when this function is available and the DPI is
known. Otherwise keep using SystemParametersInfo.

MaartenBent committed:
Use DPI Aware wxSystemParametersInfo

Require a wxWindow when requesting GetNonClientMetrics.
If no wxWindow is known, use wxTheApp->GetTopWindow().

vadz commented on src/generic/editlbox.cpp:
```diff
@@ -58,9 +58,9 @@ class CleverListCtrl : public wxListCtrl
          int w = GetSize().x;
 #ifdef __WXMSW__
-         w -= wxSystemSettings::GetMetric(wxSYS_VSCROLL_X) + 6;
+         w -= wxSystemSettings::GetMetric(wxSYS_VSCROLL_X, this) + 6;
```
Just a note (i.e. I don't say that anything needs to be changed here in this PR): we should use `FromDIP(6)` here.

MaartenBent replied:
I'll have to test this. It might be just an offset between the scrollbar and column which can be the same for any DPI.

vadz commented on src/generic/scrlwing.cpp:
```diff
@@ -1554,10 +1554,10 @@ wxSize wxScrolledT_Helper::FilterBestSize(const wxWindow *win,
 
         if ( ppuX > 0 )
-            best.x = minSize.x + wxSystemSettings::GetMetric(wxSYS_VSCROLL_X);
+            best.x = minSize.x + wxSystemSettings::GetMetric(wxSYS_VSCROLL_X, const_cast<wxWindow*>(win));
```
I think it might be better to take `const wxWindow*` parameter as querying a metric doesn't really modify the window, conceptually. And this would avoid the casts here.

MaartenBent replied:
Done.

vadz commented on src/msw/dcscreen.cpp:
```diff
@@ -41,7 +41,7 @@ wxScreenDCImpl::wxScreenDCImpl( wxScreenDC *owner ) :
 {
     if ( width )
-        *width = ::GetSystemMetrics(SM_CXVIRTUALSCREEN);
+        *width = wxGetSystemMetrics(SM_CXVIRTUALSCREEN);
```
Not sure if this is the right thing to do... Arguably, `wxScreenDC` should always use the primary monitor DPI and not the DPI of whichever monitor the application TLW is displayed on, as it's always associated with it (isn't it?).

This is not very important anyhow, of course, `wxScreenDC` is pretty useless with multiple monitors anyhow, but it might be better to just avoid changing it.

MaartenBent replied:
Reverted

vadz commented on src/msw/display.cpp:
```diff
@@ -56,7 +56,7 @@ class wxDisplayImplSingleMSW : public wxDisplayImplSingle
     {
         RECT rc;
-        SystemParametersInfo(SPI_GETWORKAREA, 0, &rc, 0);
+        wxSystemParametersInfo(SPI_GETWORKAREA, 0, &rc, 0);
```
As with `wxScreenDC` above, I think we should avoid changing this code and let it continue use the primary monitor.

vadz commented on src/msw/menuitem.cpp:
```diff
@@ -366,12 +366,12 @@ void MenuDrawData::Init()
 
         CheckMargin.cxLeftWidth =
-        CheckMargin.cxRightWidth  = ::GetSystemMetrics(SM_CXEDGE);
+        CheckMargin.cxRightWidth  = wxGetSystemMetrics(SM_CXEDGE);
```
We really need to get the actual window here somehow (i.e. `wxMenu::GetWindow()`) because using the wrong one here could easily result in drawing in the wrong size.

MaartenBent replied:
That's not possible here, as this class (`MenuDrawData`) is not associated with a menu or window. It should be taken care of in the pm-dpi PR.

vadz commented on src/msw/settings.cpp:
```diff
@@ -120,7 +120,7 @@ wxColour wxSystemSettingsNative::GetColour(wxSystemColour index)
         {
             BOOL isFlat ;
-            if ( SystemParametersInfo( SPI_GETFLATMENU , 0 ,&isFlat, 0 ) )
+            if ( wxSystemParametersInfo( SPI_GETFLATMENU , 0 ,&isFlat, 0 ) )
```
This shouldn't depend on the monitor, so arguably this shouldn't be changed neither.

vadz commented on src/msw/settings.cpp:
```diff
@@ -316,7 +316,7 @@ bool wxSystemSettingsNative::HasFeature(wxSystemFeature index)
 
         case wxSYS_TABLET_PRESENT:
-            return ::GetSystemMetrics(SM_TABLETPC) != 0;
+            return wxGetSystemMetrics(SM_TABLETPC) != 0;
```
This could be left alone too.

vadz commented on src/msw/toolbar.cpp:
```diff
@@ -581,11 +581,11 @@ wxSize wxToolBar::DoGetBestSize() const
     if ( IsVertical() )
     {
-        sizeBest.x = sizeTool.x + 2 * ::GetSystemMetrics(SM_CXBORDER);
+        sizeBest.x = sizeTool.x + 2 * wxGetSystemMetrics(SM_CXBORDER);
```
Why don't we use either `this` or `m_parent` here (and elsewhere in this file) as window parameter?

MaartenBent replied:
Thanks for noticing. Forgot to add them. This code was added after the pm-dpi PR.

vadz commented on src/msw/window.cpp:
```diff
@@ -4705,6 +4705,89 @@ wxWindowMSW::MSWOnMeasureItem(int id, WXMEASUREITEMSTRUCT *itemStruct)
+    if ( tlw )
+    {
+        wxDynamicLibrary dllUser32;
+        if ( dllUser32.Load(wxS("User32.dll"), wxDL_VERBATIM | wxDL_QUIET) )
```
I think we should use `wxLoadedDLL` here.

Also, this is a really minor nitpick, but why capitalize `User32`, we don't do it in any of the other places (including one in this file).

vadz commented on src/msw/window.cpp:
```diff
@@ -4705,6 +4705,89 @@ wxWindowMSW::MSWOnMeasureItem(int id, WXMEASUREITEMSTRUCT *itemStruct)
+        if ( dllUser32.Load(wxS("User32.dll"), wxDL_VERBATIM | wxDL_QUIET) )
+        {
+            typedef int(WINAPI *GetSystemMetricsForDpi_t)(int nIndex, UINT dpi);
+            GetSystemMetricsForDpi_t wxDL_INIT_FUNC(pfn, GetSystemMetricsForDpi, dllUser32);
```
I think this ought to be static, surely we don't want to resolve this function on every call.

vadz commented on src/msw/settings.cpp:
```diff
@@ -337,7 +337,7 @@ extern wxFont wxGetCCDefaultFont()
     // GUI font
     LOGFONT lf;
-    if ( ::SystemParametersInfo
+    if ( wxSystemParametersInfo
```
We do need the `window` parameter here however, but we don't have it in the places from which `wxGetCCDefaultFont()` is called :-( This is a problem because we're definitely going to end up with a font of the wrong size.

vadz commented on src/msw/window.cpp:
```diff
@@ -4705,6 +4705,89 @@ wxWindowMSW::MSWOnMeasureItem(int id, WXMEASUREITEMSTRUCT *itemStruct)
+        return ::SystemParametersInfo(uiAction, uiParam, pvParam, fWinIni);
+    }
+
+    wxTopLevelWindow* tlw = NULL;
```
This code should be extracted into a separate function to avoid duplicating it between this one and `wxGetSystemMetrics()` or maybe even to call it explicitly (see the global review comment).

vadz commented on src/msw/window.cpp:
```diff
@@ -4705,6 +4705,89 @@ wxWindowMSW::MSWOnMeasureItem(int id, WXMEASUREITEMSTRUCT *itemStruct)
+
+    if ( tlw )
+    {
+        wxDynamicLibrary dllUser32;
```
Same remarks as above apply to this one too, i.e. let's resolve the function only once etc.

vadz reviewed (commented):
First of all, thanks a lot for extracting this into its own PR, it made the changes much easier to review.

I do have some suggestions for improvement now that I could actually look at it and the main one is that I think that the argument of the new functions should _not_ be optional, i.e. shouldn't default to `NULL`. Instead, I'd rather explicitly use `wxApp::GetTopWindow()` (maybe add a `wxGetAppTLW()` helper that would be safe to use in all cases, i.e. check for `wxTheApp` being non-null, and also apply dynamic cast inside it) in the places where no more appropriate window is available, to make it easier to search for them and fix them later. What do you think of doing it like this? Hopefully it shouldn't require too much effort to change this, but it will hopefully prevent people from lazily calling these functions without any window parameter in the future, which is IMO an important advantage.

I'm going to look at the font-related changes next, so maybe the other comment would be more appropriate there, but I also think we need to do something about DPI-independent and actual fonts. I.e. `wxGetCCDefaultFont()` simply can't return the font of the correct height right now, but silently using the wrong height is really not great...

Thanks again for your work on this!

MaartenBent committed:
Remove static reference to NONCLIENTMETRICS

These metrics can change when the DPI of a Window changes, so we can not keep a
static reference.

According to documentation, the second parameter (uiParam) should be set to the
size of the NONCLIENTMETRICS object.

MaartenBent commented:
> wxGetCCDefaultFont() simply can't return the font of the correct height right now, but silently using the wrong height is really not great...

The pointsize of the `wxFont` will be correct. Once the font is being used it will be adjusted to the active DPI.

MaartenBent committed:
Use const wxWindow in wxSystemSettings::GetMetric

vadz commented:
Thanks for the update!
